### PR TITLE
[CSL-870] Use NAT network-transport-tcp; wallet and smart-generator are unaddressable

### PR DESCRIPTION
**Pull request by avieth**
Included here:

1. Fix up the log warning timeout. There was a possibility that the warning printing thread was never killed. I'm not suggesting that the warnings we see in the logs of massive recv times are due to this, it's just something I noticed that *could* cause misleading logs.
2. Include network-transport-tcp bug fix which sometimes caused connections to be lost prematurely.
3. Make the wallet and smart-generator `Unaddressable`. They will not claim `127.0.0.1` as their host. However, they still join Kademlia, bound to `0.0.0.0:0`. The peers will get some random ip/port and keep on trying to connect to it, to no avail. This should be benign (that's what was happening anyway, it's not introduced by this patch).

### CODE REVIEW

**Comment by avieth** [comment]
I've also cherry-picked these onto the qanet branch https://github.com/input-output-hk/cardano-sl/tree/cardano-sl-0.4-qanet-nat

Today in some logs that @georgeee sent me I found that towards the end of the run, node 0 spent 600 seconds trying to send to `127.0.0.1:12100`. I spent some time trying to figure out how that could possibly happen. A time-warp-nt send goes straight to a network-transport send, with no blocking code around it. A network-transport-tcp send should return as soon as the data is in the kernel buffer. I couldn't see a way that the send would possibly block for 600 seconds. If there were a network-transport-tcp server listening at `127.0.0.1:12100` (there isn't) then it would surely receive some data and make room in the send buffer before 600 seconds. If there were no server there then we wouldn't even get that far; an exception would be thrown before trying to send.

Hopefully this was just a case of the warning thread not being killed (its parent thread being killed before it can call `killThread`, which is interruptible i.e. can be killed even when in masked state). Though I can't explain how that would happen either (where is an exception ever thrown to a thread that is conversing with a peer?).

Waiting on a `recv` for a very long time is a different matter. However, there's a timeout of 5 seconds for `send` and `recv`, so obviously we shouldn't ever see a warning of more than `5s`. The warning thread escaping because the `killThread` is interrupted also explains this. Or perhaps `execWithTimeLimit` has a subtle bug? Could its use of `cancel` potentially be interrupted or itself block indefinitely (because it's trying to cancel some uninterruptible operation)?

File: infra/Pos/Util/TimeLimit.hs [hunk 1]
```diff
@@ -38,17 +38,25 @@ data WaitingDelta
...
 -- | Constraint for something that can be logged in parallel with other action.
-type CanLogInParallel m = (Mockable Delay m, Mockable Fork m, WithLogger m, Mockable Bracket m)
+type CanLogInParallel m = (Mockable Delay m, Mockable Async m, WithLogger m)
 
 
 -- | Run action and print warning if it takes more time than expected.
 logWarningLongAction :: CanLogInParallel m => WaitingDelta -> Text -> m a -> m a
 logWarningLongAction delta actionTag action =
-    bracket (fork $ waitAndWarn delta) onFinish (const action)
+    -- Previous implementation was
```
Comment (gromakovsky) [inline infra/Pos/Util/TimeLimit.hs]: @ChShersh take a look, it may be interesting for you as you were writing it.

**Comment by domenkozar** [comment]
@avieth to fix tests please rebase master

File: src/Pos/Launcher/Param.hs [hunk 2]
```diff
@@ -28,7 +28,7 @@ data LoggingParams = LoggingParams
 
 -- | Contains basic & networking parameters for running node.
 data BaseParams = BaseParams
-    { bpIpPort             :: !NetworkAddress -- ^ Port to run on
+    { bpIpPort             :: !(Maybe NetworkAddress) -- ^ Port to run on
```
Comment (volhovM) [inline src/Pos/Launcher/Param.hs]: Isn't it a somewhat counterintuitive naming?

File: src/Pos/Launcher/Runner.hs [hunk 3]
```diff
@@ -451,10 +453,14 @@ createTransport ip port = do
...
 bracketTransport BaseParams {..} =
     bracket
-        (withLog $ createTransport (BS8.unpack $ fst bpIpPort) (snd bpIpPort))
+        (withLog $ createTransport (maybe TCP.Unaddressable TCP.Addressable addrInfo))
         (liftIO . closeTransport)
   where
     withLog = usingLoggerName $ lpRunnerTag bpLoggingParams
+    addrInfo = do
+        host <- fmap (BS8.unpack . fst) bpIpPort
+        port <- fmap (show . snd) bpIpPort
```
Comment (volhovM) [inline src/Pos/Launcher/Runner.hs]: More concise:
```
(host, port) <- bimap BS8.unpack show <$> bpIpPort
```

File: src/Pos/Worker.hs [hunk 4]
```diff
@@ -35,7 +35,10 @@ allWorkers
 allWorkers = mconcatPair
     [
       -- Only workers of "onNewSlot" type
+
+      -- TODO cannot have this DHT worker here. It assumes Kademlia.
```
Comment (volhovM) [inline src/Pos/Worker.hs]: What does it mean? I guess `WorkMode` constraint implies we have kademlia.

File: src/Pos/Launcher/Param.hs (hunk 2 above)
Comment (avieth) [inline src/Pos/Launcher/Param.hs]: How about I change it to `bpNetworkAddress`?

File: src/Pos/Worker.hs (hunk 4 above)
Comment (avieth) [inline src/Pos/Worker.hs]: Yes `WorkMode` assumes `MonadDHT` and `MonadDHT` is basically fixed to Kademlia. I'm working on abstracting the discovery mechanism. It's important that we can swap Kademlia out in favour of, for instance, a fixed set of peers. Helps with testing and profiling. So my idea here is to remove these listeners from `allListeners` and instead have them included by some `Pos.Launcher` module which chooses Kademlia.

File: src/Pos/Launcher/Param.hs (hunk 2 above)
Comment (avieth) [inline src/Pos/Launcher/Param.hs]: Renamed all `ipPort` options to `networkAddress`

File: src/Pos/Launcher/Runner.hs [hunk 5]
```diff
@@ -422,7 +422,9 @@ bracketDHTInstance BaseParams {..} action = bracket acquire release action
     instConfig =
         KademliaDHTInstanceConfig
         { kdcKey = bpDHTKey
-        , kdcHost = maybe "0.0.0.0" fst bpNetworkAddress
+        --, kdcHost = maybe "0.0.0.0" fst bpNetworkAddress
+        , kdcHost = "0.0.0.0" -- Amazon servers have different ifconfig (local) ip and public ip
+                              -- so we can't bind on public directly.
```
Comment (avieth) [inline src/Pos/Launcher/Runner.hs]: The right solution here is to have a separate Kademlia configuration. I have this in my branch work to abstract the discovery mechanism.

Comment (avieth) [inline src/Pos/Launcher/Runner.hs]: But then again, the system assumes that if you use Kademlia, then the address of the UDP socket that Kademlia uses is the same as that of the network-transport-tcp public address. We should build this in and make it impossible to violate.

**Comment by domenkozar** [comment]
Windows appveyor fails because (I know ...) too long paths.

I recommend shortening `cardano-wallet-web-api-swagger` to `cardano-swagger` or something similar.

**Comment by volhovM** [comment]
@domenkozar I've renamed it to `cardano-swagger`, appveyor still fails (with the same error).

**Comment by domenkozar** [comment]
I don't see how this PR causes paths to be too long. I've cleared appveyor cache and restarted the build to make sure it's not transient.

**Comment by domenkozar** [comment]
So here's the a new hypothesis: long path is coming from one of the bump packages: `kademlia`, `time-warp-nt` or `network-transport-tcp`

**Comment by avieth** [comment]
> So here's the a new hypothesis: long path is coming from one of the bumped packages: kademlia, time-warp-nt or network-transport-tcp

Had a look and I didn't see anything in kademlia or time-warp-nt that could introduce a 260 character path (that's the issue right?). However the new network-transport-tcp rev includes a new dependency: `uuid`. Perhaps its longest path `uuid/uuid-types/Data/UUID/Types/Internal/Builder.hs` pushes it over the limit?

**Comment by domenkozar** [comment]
The fix is in https://github.com/input-output-hk/cardano-sl/pull/236

**Pull request merged**